libs/ml/examples/boston_housing_distributed_learning/main.cpp

```diff
@@ -0,0 +1,220 @@
+  client_params.learning_rate                   = static_cast<DataType>(.0001f);
+  float                       test_set_ratio    = 0.00f;
+  std::shared_ptr<std::mutex> console_mutex_ptr = std::make_shared<std::mutex>();
+
+  // Load data
+  TensorType data_tensor  = fetch::ml::dataloaders::ReadCSV<TensorType>(av[1]).Transpose();
+  TensorType label_tensor = fetch::ml::dataloaders::ReadCSV<TensorType>(av[2]).Transpose();
+
+  SizeType data_size = data_tensor.shape().at(1);
+
+  // Split data for each client
+  std::vector<SizeType> splitting_points;
+
+  SizeType client_data_size         = data_size / number_of_clients;
+  SizeType index                    = 0;
+  SizeType current_client_data_size = client_data_size;
+  for (SizeType i = 0; i < number_of_clients; i++)
+  {
+    if (i == number_of_clients - 1)
+    {
+      current_client_data_size = data_size - index;
+    }
+    splitting_points.push_back(current_client_data_size);
+    index += client_data_size;
+  }
+
+  std::vector<TensorType> data_tensors  = TensorType::Split(data_tensor, splitting_points, 1);
+  std::vector<TensorType> label_tensors = TensorType::Split(label_tensor, splitting_points, 1);
+
+  // Create coordinator
+  std::shared_ptr<Coordinator<TensorType>> coordinator =
+      std::make_shared<Coordinator<TensorType>>(coord_params);
+
+  std::cout << "FETCH Distributed BOSTON HOUSING Demo" << std::endl;
+
+  std::vector<std::shared_ptr<TrainingClient<TensorType>>> clients(number_of_clients);
+  for (SizeType i{0}; i < number_of_clients; ++i)
+  {
+
+    // Instantiate NUMBER_OF_CLIENTS clients
+    clients[i] = MakeClient(i, client_params, data_tensors.at(i), label_tensors.at(i),
+                            test_set_ratio, console_mutex_ptr);
+    // TODO(1597): Replace ID with something more sensible
+  }
+
+  // Give list of clients to coordinator
+  coordinator->SetClientsList(clients);
+
+  for (SizeType i{0}; i < number_of_clients; ++i)
+  {
+    // Give each client pointer to coordinator
+    clients[i]->SetCoordinator(coordinator);
+  }
+
+  // Main loop
+  for (SizeType it{0}; it < number_of_rounds; ++it)
+  {
+    // Start all clients
+    coordinator->Reset();
+    std::cout << "ROUND : " << it << "\t";
+    std::list<std::thread> threads;
+    for (auto &c : clients)
+    {
+      threads.emplace_back([&c] { c->Run(); });
+    }
+
+    // Wait for everyone to finish
+    for (auto &t : threads)
+    {
+      t.join();
+    }
+
+    std::cout << "Test losses:";
+    for (auto &c : clients)
+    {
+      std::cout << "\t" << static_cast<double>(Test(c->GetModel(), data_tensor, label_tensor));
+    }
+    std::cout << std::endl;
+
+    if (coordinator->GetMode() == CoordinatorMode::ASYNCHRONOUS)
+    {
+      continue;
+    }
+
+    // Synchronize weights by giving all clients average of all client's weights
+    VectorTensorType new_weights = clients[0]->GetWeights();
+
+    // Sum all weights
+    for (SizeType i{1}; i < number_of_clients; ++i)
```

Comment:
separate functions for summing all weights and dividing weights can also make the main loop clearer